File: src/main/java/cm/xd/oopdate/query/Query.java
package cm.xd.oopdate.query;

import lombok.Data;

@Data
public class Query {
    private String castType;
    private QueryType queryType;
    private String name;
    private Object value;

    private static Query getQuery(String name, Object value) {
        Query q = new Query();
        q.setName(name);
        q.setValue(value);
        return q;
    }

    public static Query identity(String name, Object value) {
        Query q = getQuery(name, value);
        q.setQueryType(QueryType.IDENTITY);
        return q;
    }

    public static Query data(String name, Object value) {
        Query q = getQuery(name, value);
        q.setQueryType(QueryType.DATA);
        return q;
    }

    public Query cast(String castType) {
        this.castType = castType;
        return this;
    }

    /**
     * @deprecated use string method instead.
     * @param castType deprecated, CastType to use
     * @return Query
     */
    @Deprecated
    public Query cast(CastType castType) {
        this.castType = castType.toString().toLowerCase();
        return this;
    }
}

File: src/main/kotlin/cm/xd/oopdate/query/Query.kt
package cm.xd.oopdate.query;

class Query {
    var castType: String? = null
    var queryType: QueryType? = null
    var name: String? = null
    var value: kotlin.Any? = null

    companion object {

        @JvmStatic
        private fun getQuery(name: String, value: kotlin.Any): Query {
            val q = Query()
            q.name = name
            q.value = value
            return q
        }

        @JvmStatic
        fun identity(name: String, value: kotlin.Any): Query
        {
            val q = getQuery(name, value)
            q.queryType = QueryType.IDENTITY
            return q
        }

        @JvmStatic
        fun data (name: String, value: kotlin.Any): Query
        {
            val q = getQuery(name, value)
            q.queryType = QueryType.DATA
            return q
        }
    }

    fun cast(castType: String): Query {
        this.castType = castType
        return this
    }

    /**
     * @deprecated use string method instead.
     * @param castType deprecated, CastType to use
     * @return Query
     */
    @Deprecated(message = "Use cast(String) instead")
    fun cast(castType: CastType): Query {
        this.castType = castType.toString().toLowerCase()
        return this
    }
}
